### PR TITLE
Regional strip unpacker: support for legacy modes and bugfix for virgin raw

### PR DESCRIPTION
**Pull request by pieterdavid**
These are two fixes, adapted from https://github.com/CMS-HIN-dilepton/cmssw/commit/2742728c4ae203936ebbd31d50934d9b1938d0cb?w=1 (prepared some time ago but never submitted), to make the regional unpacker (used in some HLT paths) work also with legacy readout modes, and with virgin raw data (the corresponding code in the full unpacker is in https://github.com/cms-sw/cmssw/blob/master/EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.cc#L436-L440 ). This scenario is not considered by any relval, as far as I know, so the HI group should confirm that this works as expected.
As I understood, there is no plan to use virgin raw for the upcoming HI run, but we are submitting this bugfix to have it in place just in case.

CC: @icali (I cannot find Andre on github, could you ping him?) @alesaggio

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@pieterdavid 
please remind me if there are test configs that test the alternative configurations/modes? (unit tests would be even better)

File: RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc [hunk 1]
```diff
@@ -181,8 +184,14 @@ namespace {
...
   };
-  
-  
+
+  uint16_t physicalToReadoutOrder( uint16_t physical_order )
+  {
+    auto phys_float = static_cast<float>(physical_order);
+    return ( 4*(static_cast<uint16_t>(phys_float/ 8.0)%4)
+           +    static_cast<uint16_t>(phys_float/32.0)
+           + 16*(physical_order%8) );
```
[the remark this answers is not in the document]
Comment (pieterdavid) [reply RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromRawProducer.cc]: Indeed, that would be more efficient, and more readable as well (I copied the code from https://github.com/cms-sw/cmssw/blob/9b1f3a74c13aaf742d172fb68afec34f6fecc8d9/EventFilter/SiStripRawToDigi/plugins/SiStripRawToDigiUnpacker.h#L155 and cleaned it up a bit but didn't look in detail before).
The calling code does `%128` first, and then applies, depending on whether the 8th bit was 0 or 1, `*2` or `*2+1)` on the result - effectively swapping with the 7 considered here (the number is known not to be larger than 255), so I replaced that by bit operations as well (and removed the separate method).

**Comment by pieterdavid** [comment]
@slava77 no, this is not specifically tested by relvals or unit tests, that's why I asked for confirmation from the people working on heavy ion triggers who found this bug. The corresponding code in the full unpacker is covered by tests, and I want to reduce the code duplication between the two as part of the refactoring we will need to do for https://github.com/cms-sw/cmssw/issues/24825 , such that we don't need specific tests for the unpacking part here.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24851	91d3de7 
- updates are in line with the PR description: adding appropriate unpacking calls for the new packing modes in the clustering directly done from RAW with unpacking called in situ. This affects only HLT setup and is in fact not expected to be used in the production environment.
- jenkins tests pass and the comparisons with the baseline show no differences.

**Comment by slava77** [comment]
assign hlt

**Comment by slava77** [comment]
I've requested the hlt signature to formalize the request made in the PR description

**Comment by Martin-Grunewald** [comment]
This code has neither a fillDescriptions nor adds the parameter to any cfi file.
That's no good for ConfDB parsing!
And does the default of "FALSE" reproduce the old behaviour?

**Comment by pieterdavid** [comment]
@Martin-Grunewald yes, LegacyUnpacker=false reproduces the old behaviour (legacy is for old versions of the firmware, that haven't been used for a while, as far as I know).
I can add a fillDescriptions method, but since two of the parameters are for the [zero-suppression algorithms](https://github.com/cms-sw/cmssw/blob/master/RecoLocalTracker/SiStripZeroSuppression/python/DefaultAlgorithms_cff.py) and [clusterizer](https://github.com/cms-sw/cmssw/blob/master/RecoLocalTracker/SiStripClusterizer/python/DefaultClusterizer_cff.py) configuration, to avoid duplication I should delegate that part to a fillDescriptions method of the corresponding factories, which are then also used by the other modules that use these - I think this would be a good idea in general, but it may be beyond the scope of this PR.

**Comment by Martin-Grunewald** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**